src/go/k8s/internal/client/factory_test.go

```diff
@@ -0,0 +1,242 @@
+	"fmt"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/redpanda-data/helm-charts/pkg/helm"
+	"github.com/redpanda-data/helm-charts/pkg/kube"
+	"github.com/redpanda-data/helm-charts/pkg/testutil"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/redpanda-data/redpanda-operator/src/go/k8s/pkg/k3d"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kadm"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var chartVersion = ""
+
+func ensureMapAndSetValue(values map[string]any, name, key string, value any) {
+	if v, ok := values[name]; ok {
+		m := v.(map[string]any)
+		m[key] = value
+		values[name] = m
+
+		return
+	}
+
+	values[name] = map[string]any{
+		key: value,
+	}
+}
+
+type fakeObject struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	kafkaSpec *redpandav1alpha2.KafkaAPISpec
+}
+
+func wrapSpec(name string, spec *redpandav1alpha2.KafkaAPISpec) *fakeObject {
+	return &fakeObject{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: name,
+		},
+		kafkaSpec: spec,
+	}
+}
+
+func (f *fakeObject) GetKafkaAPISpec() *redpandav1alpha2.KafkaAPISpec {
+	return f.kafkaSpec
+}
+
+func (f *fakeObject) DeepCopyObject() runtime.Object {
+	return f
+}
+
+func TestClientFactory(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping factory tests in short mode")
+	}
+
+	var suffix atomic.Int32
+
+	ctx := context.Background()
+	cluster, err := k3d.NewCluster(t.Name())
+	require.NoError(t, err)
+	t.Logf("created cluster %T %q", cluster, cluster.Name)
+
+	t.Cleanup(func() {
+		if testutil.Retain() {
+			t.Logf("retain flag is set; not deleting cluster %q", cluster.Name)
+			return
+		}
+		t.Logf("Deleting cluster %q", cluster.Name)
+		require.NoError(t, cluster.Cleanup())
+	})
+
+	restcfg := cluster.RESTConfig()
+
+	s := runtime.NewScheme()
+	require.NoError(t, clientgoscheme.AddToScheme(s))
+	require.NoError(t, redpandav1alpha2.AddToScheme(s))
+	require.NoError(t, redpandav1alpha1.AddToScheme(s))
+	kubeClient, err := client.New(restcfg, client.Options{Scheme: s})
+	require.NoError(t, err)
+
+	helmClient, err := helm.New(helm.Options{
+		KubeConfig: restcfg,
+	})
+	require.NoError(t, err)
+	require.NoError(t, helmClient.RepoAdd(ctx, "jetstack", "https://charts.jetstack.io"))
+	require.NoError(t, helmClient.RepoAdd(ctx, "redpandadata", "https://charts.redpanda.com"))
+
+	_, err = helmClient.Install(ctx, "jetstack/cert-manager", helm.InstallOptions{
+		CreateNamespace: true,
+		Name:            "cert-manager",
+		Namespace:       "cert-manager",
+		Values: map[string]any{
+			"crds": map[string]any{
+				"enabled": true,
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	factory, err := NewClientFactory(restcfg)
```

Comment:
NIT:

I saw the following problem when running the test
```
[controller-runtime] log.SetLogger(...) was never called; logs will not be displayed.
Detected at:
	>  goroutine 13 [running]:
	>  runtime/debug.Stack()
	>  	/opt/homebrew/opt/go/libexec/src/runtime/debug/stack.go:24 +0x64
	>  sigs.k8s.io/controller-runtime/pkg/log.eventuallyFulfillRoot()
	>  	/Users/rafalkorepta/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/log/log.go:60 +0xf4
	>  sigs.k8s.io/controller-runtime/pkg/log.(*delegatingLogSink).WithName(0x14000408600, {0x104bd892a, 0x14})
	>  	/Users/rafalkorepta/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/log/deleg.go:147 +0x34
	>  github.com/go-logr/logr.Logger.WithName({{0x1056d2470, 0x14000408600}, 0x0}, {0x104bd892a?, 0x140003000e0?})
	>  	/Users/rafalkorepta/go/pkg/mod/github.com/go-logr/logr@v1.4.1/logr.go:345 +0x40
	>  sigs.k8s.io/controller-runtime/pkg/client.newClient(0x14000561ab8?, {0x0, 0x140003000e0, {0x0, 0x0}, 0x0, {0x0, 0x0}, 0x0})
	>  	/Users/rafalkorepta/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/client/client.go:129 +0xb4
	>  sigs.k8s.io/controller-runtime/pkg/client.New(0x140003000e0?, {0x0, 0x140003000e0, {0x0, 0x0}, 0x0, {0x0, 0x0}, 0x0})
	>  	/Users/rafalkorepta/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/client/client.go:110 +0x54
	>  github.com/redpanda-data/redpanda-operator/src/go/k8s/internal/client.NewClientFactory(0x14000816008)
	>  	/Users/rafalkorepta/workspace/redpanda/redpanda-operator/src/go/k8s/internal/client/factory.go:82 +0xa8
	>  github.com/redpanda-data/redpanda-operator/src/go/k8s/internal/client.TestClientFactory(0x1400015fba0)
	>  	/Users/rafalkorepta/workspace/redpanda/redpanda-operator/src/go/k8s/internal/client/factory_test.go:115 +0x510
	>  testing.tRunner(0x1400015fba0, 0x105695e30)
	>  	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0xec
	>  created by testing.(*T).Run in goroutine 1
	>  	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x318
```

Reply:
Yeah, so I'm not entirely sure how we want to deal with that, namely because it comes from unfortunate global logger delegation in the controller-runtime code.

I believe you can suppress the log calls on a per-cliient basis by initializing a client with something like passing:

```go
client.Options{WarningHandler: client.WarningHandlerOptions{SuppressWarnings: true}}
```

into a `client.New` call... however, I doubt we want to do that everywhere just for the sake of suppressing the errors in tests. Alternatively, I could also potentially add a parameter to the constructor for `ClientFactory` so you'd need to call `NewClientFactory(restcfg, kubeClient)` instead of just `NewClientFactory(restcfg)` as it is now, but, you'd need to ensure that the client passed to it had already been initialized properly with a scheme that has both core and our CRDs registered with it, which feels pretty leaky and unnecessary.

The last option would be calling the global `sigs.k8s.io/controller-runtime@v0.17.2/pkg/log.SetLogger` as indicated in the warning which should be fine in terms of thread safety, but would affect all internal runtime log calls for every running test, which also seems really, really weird.

I'm somewhat inclined just to leave it as is since I think it may just log that error once across all tests even outside of this file when we initialize a controller-runtime `client.Client`, but I'm also open to suggestions.

Reply:
Agree

Reply:
I think the aforementioned context logging helpers can work around this. Here's what I did in the PVC unbinder:
```go
	logger := testr.New(t).V(0)
	log.SetLogger(logger)
	ctx = log.IntoContext(ctx, logger)
```

Reply:
So, in this case, the contextual logger wouldn't really come into play anyway because you'd still need to call the global `log.SetLogger`, which seems really odd in that you're mutating a _**global**_ internal logging instance that every code path using a controller-runtime `Client` hits. While it's threadsafe and all, it just feels really odd to do a call that has global side effects for every other test in a single test (i.e. there are some other tests running this same code, if they ran first before this test gets run we wouldn't see the logging warning stack trace).

That said, the contextual logger may be a good option for passing along a logger to the kgo/rpadmin client initialization routines (though I don't think it'd help for a controller-runtime Client initialization)!

Reply:
Wound up using `IntoContext` within the topic controller to pass along the scoped logger. Also, just to silence the warnings on any tests that may use this, through a `log.SetLogger(logr.Discard())` in an `init` function in the top of the test file just so we do it once in this package...